### PR TITLE
Better support for "audience" / simplified Lemmy processing

### LINKED ISSUE #13173: OpenWebAuth login on remote server fails if Friendica server does not have a contact
- [X] I have searched open and closed issues for duplicates

### Bug Description

When trying to perform an OpenWebAuth login on e.g. a Hubzilla server through a Friendica login, the login fails if the Friendica server does not have a contact on the Hubzilla server. The OWA login is not even started because the Friendica server bails out before.

### Steps to Reproduce

1.  Log in on Friendica server
2.  On a Hubzilla server, go to remote authentication on /rmagic
3.  Enter the friendica handle and press 'Authenticate'

Actual Result:

The Hubzilla server returns to the login page, no user is logged in

Expected Result:

The Friendica user should be logged in on the Hubzilla server as visitor

### Screenshots

When the OWA-based login works, then the Hubzilla avatar in the left upper corner changes to the Friendica avatar. If not, it doesn't.

### Platform Info

Friendica Version: Friendica 'Giant Rhubarb' 2023.09-dev - 1518

### Troubleshooting

I see these Friendica log messages:
```
{"file":"Contact.php","line":1335,"function":"getIdForURL","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
2023-05-26T05:52:57Z index [WARNING]: E_WARNING: Undefined array key "gsid" {"code":2,"message":"Undefined array key \"gsid\"","file":"/srv/friendica/src/Module/Magic.php","line":113} - {"file":"Magic.php","line":113,"function":"rawContent","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
2023-05-26T05:52:57Z index [WARNING]: E_WARNING: Undefined array key "gsid" {"code":2,"message":"Undefined array key \"gsid\"","file":"/srv/friendica/src/Module/Magic.php","line":115} - {"file":"Magic.php","line":115,"function":"rawContent","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
2023-05-26T05:52:57Z index [NOTICE]: Server not found, redirecting to destination. {"gsid":null,"dest":"https://hzsandbox2.eenoog.org/"} - {"file":"Magic.php","line":115,"function":"rawContent","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
```

I think this is due to this code in src/Module/Magic.php:

		if ($addr ?: $dest) {
			$contact = Contact::getByURL($addr ?: $dest);
		}

...

		$gserver = $this->dba->selectFirst('gserver', ['url'], ['id' => $contact['gsid']]);
		if (empty($gserver)) {
			$this->logger->notice('Server not found, redirecting to destination.', ['gsid' => $contact['gsid'], 'dest' => $dest]);
			System::externalRedirect($dest);
		}

Indeed, it is possible to reach this code if the contact is empty but in case of OWA, that's fine: it is not required for OWA that the Friendica server knows an ActivityPub actor on Hubzilla.

In my server, I have these entries in the 'gserver' table:

> MariaDB [friendica]> select id,url from gserver;
> +----+-------------------------------+
> | id | url                           |
> +----+-------------------------------+
> |  1 | https://friendica.eenoog.org  |
> |  2 | https://hub.eenoog.org        |
> |  3 | https://stroom.eenoog.org     |
> |  4 | https://hzsandbox2.eenoog.org |
> |  5 | http://f.lapo.it              |
> |  6 | https://sandpix.eenoog.org    |
> |  7 | https://fd.winklerfamilie.eu  |
> |  8 | https://hub.somaton.com       |
> +----+-------------------------------+

In my case, the Hubzilla server is 'hzsandbox2.eenoog.org', and I have no contact on that server:
> MariaDB [friendica]> select id,uid,nick,baseurl from contact;
> +----+-----+-----------+------------------------------+
> | id | uid | nick      | baseurl                      |
> +----+-----+-----------+------------------------------+
> |  0 |   0 |           |                              |
> |  2 |   2 | pascal    |                              |
> |  3 |   0 | pascal    | https://friendica.eenoog.org |
> |  4 |   0 | friendica | https://friendica.eenoog.org |
> |  6 |   0 | sys       | http://stroom.eenoog.org     |
> |  7 |   0 | pascal    | http://stroom.eenoog.org     |
> |  8 |   2 | pascal    | http://stroom.eenoog.org     |
> |  9 |   0 | pascal    | https://stroom.eenoog.org    |
> | 10 |   0 | friendica | https://f.lapo.it            |
> | 11 |   0 | abanink   | http://sandpix.eenoog.org    |
> | 12 |   0 | friendica | https://fd.winklerfamilie.eu |
> +----+-----+-----------+------------------------------+

which explains the behaviour. 
In my view, it's a bug but I'm not enough known with Friendica internals to propose a fix.

### PR DESCRIPTION
**Pull request by annando**
This PR does contain a mixture of different tasks, since working on the central task took a long time. With these changes a lot of issues with Lemmy should be fixed now. Posts (and comments) to Lemmy will now be distributed only to the server, not the own followers.

To explain: Lemmy is using the "audience" field for signalling that it is a group post. We introduced that field for our own groups as well (this PR here also contains some improvements there as well). So we now have a close look on that field and process that in the best way.

Also this PR does contain some fixes that occur in between. For example Bluesky-URIs are no HTTP links, so clicking on the don't make sense. So we now automatically display a link to the contact "alias" if the URL is no HTTP link. Also some parts of the code are reformatted as well.

### CODE REVIEW

File: src/Content/Widget/VCard.php [hunk 1]
```diff
@@ -50,9 +51,15 @@ public static function getHTML(array $contact): string
 			Logger::warning('Incomplete contact', ['contact' => $contact ?? [], 'callstack' => System::callstack(20)]);
 		}
 
+		if (!Network::isValidHttpUrl($contact['url']) && Network::isValidHttpUrl($contact['alias'])) {
+			$url = $contact['alias'];
+		} else {
+			$url = $contact['url'];
```
Comment (MrPetovan) [inline src/Content/Widget/VCard.php]: Please rename the variable to `$contact_url` for better visibility.
```suggestion
			$contact_url = $contact['alias'];
		} else {
			$contact_url = $contact['url'];
```

File: src/Model/Item.php [hunk 2]
```diff
@@ -3664,14 +3733,21 @@ public static function getPlink(array $item)
 
 	/**
 	 * Does the given uri-id belongs to a post that is sent as starting post to a group?
+	 * This does not apply to posts that are sent only in parallel to a group.
```
Comment (MrPetovan) [inline src/Model/Item.php]: What does this mean? Are both behaviors (parallel and not-parallel?) described somewhere?

File: src/Module/Contact.php [hunk 3]
```diff
@@ -412,20 +411,41 @@ protected function content(array $request = []): string
...
+				break;
+			case 'ignored':
+				$header .= ' - ' . DI::l10n()->t('Ignored');
+				break;
+			case 'collapsed':
+				$header .= ' - ' . DI::l10n()->t('Collapsed');
+				break;
+			case 'archived':
+				$header .= ' - ' . DI::l10n()->t('Archived');
+				break;
```
Comment (MrPetovan) [inline src/Module/Contact.php:448]: Less legible change, please revert.

File: src/Module/Contact.php [hunk 4]
```diff
@@ -412,20 +411,41 @@ protected function content(array $request = []): string
...
+				$header = DI::l10n()->t('Following');
+				break;
+			case 'mutuals':
+				$header = DI::l10n()->t('Mutual friends');
+				break;
+			case 'nothing':
+				$header = DI::l10n()->t('No relationship');
+				break;
+			default:
+				$header = DI::l10n()->t('Contacts');
```
Comment (MrPetovan) [inline src/Module/Contact.php:427]: Less legible change, please revert.

File: src/Module/Magic.php [hunk 5]
```diff
@@ -83,6 +85,8 @@ protected function rawContent(array $request = [])
 			$this->logger->debug('bdest detected', ['dest' => $dest]);
 		}
 
+		$target = $dest ?: $addr;
+
 		if ($addr ?: $dest) {
```
Comment (MrPetovan) [inline src/Module/Magic.php:90]: What are the cases where `dest` is set but not `addr`, and on the opposite side, `addr` is set but not `dest`?

**Review by MrPetovan — CHANGES_REQUESTED** [review]
Next time, please submit your formatting changes separately.

File: src/Model/Item.php (hunk 2 above)
Comment (annando) [reply src/Model/Item.php]: This is the difference between mentioning a group via `!` and `@`.

File: src/Module/Contact.php (hunk 4 above)
Comment (annando) [reply src/Module/Contact.php:427]: From my point of view I find this much more readable, I must admit.

Comment (MrPetovan) [reply src/Module/Contact.php:427]: Interestingly, PHP 8 has a structure to replace these verbose switch statements: https://www.php.net/manual/en/control-structures.match.php

File: src/Module/Magic.php (hunk 5 above)
Comment (annando) [reply src/Module/Magic.php:90]: Concerning this module: I don't know much about it. I'm just trying to fix stuff.

**Comment by annando** [comment]
I will be on holiday for the next few days, so I cannot guarantee when I will address the comments.

Unrelated to this PR: Please don't forget my PR in the addons.

File: src/Module/Magic.php (hunk 5 above)
Comment (MrPetovan) [reply src/Module/Magic.php:90]: How do you test then?

File: src/Model/Item.php (hunk 2 above)
Comment (tobiasd) [reply src/Model/Item.php]: Could you add the "@-mention" as explanation to the "parallel" (if that is meant).

File: src/Module/Magic.php (hunk 5 above)
Comment (annando) [reply src/Module/Magic.php:90]: I guess like you tested OStatus when you applied some changes :-)

Comment (MrPetovan) [reply src/Module/Magic.php:90]: So those are blind changes? I'm just trying to understand the purpose.

Comment (annando) [reply src/Module/Magic.php:90]: In fact, I made them a longer time ago and had to had a deeper look at it to know what the changes mean.

File: src/Model/Item.php (hunk 2 above)
Comment (annando) [reply src/Model/Item.php]: Changed

File: src/Module/Magic.php (hunk 5 above)
Comment (MrPetovan) [reply src/Module/Magic.php:90]: Alrighty, let's leave this for future people to figure it out.

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**